gateway/gateway.go

```diff
@@ -1491,6 +1494,38 @@ func (*HandleT) GetWebhookSourceDefName(writeKey string) (name string, ok bool)
 	return
 }
 
+// SaveErrors saves errors to the error db
+func (gateway *HandleT) SaveWebhookFailures(reqs []*model.FailedWebhookPayload) error {
+	jobs := make([]*jobsdb.JobT, 0, len(reqs))
+	for _, req := range reqs {
+		params := map[string]interface{}{
+			"source_id":   gateway.getSourceIDForWriteKey(req.WriteKey),
+			"stage":       "webhook",
+			"source_type": req.SourceType,
+			"reason":      req.Reason,
+		}
+		marshalledParams, err := json.Marshal(params)
+		if err != nil {
+			gateway.logger.Errorf("[Gateway] Failed to marshal parameters map. Parameters: %+v", params)
+			marshalledParams = []byte(`{"error": "rudder-server gateway failed to marshal params"}`)
+		}
+
+		jobs = append(jobs, &jobsdb.JobT{
+			UUID:         uuid.New(),
+			UserID:       uuid.New().String(), // Using a random userid for these failures. There is no notion of user id for these events.
+			Parameters:   marshalledParams,
+			CustomVal:    "WEBHOOK",
+			EventPayload: req.Payload,
```

Comment:
Just thinking out loud... what if the req.Payload is not a JSON but rather XML data? or some other format

Reply:
Correct, payload should be stored as a valid `json`, otherwise store would be retried twice and ultimately fail

Reply:
A gateway test scenario using an actual jobsdb would be useful here

Reply:
We validate the incoming payload to be JSON [here](https://github.com/rudderlabs/rudder-server/blob/a8abd1db5b6a9b434dd3363b3f959302ecdfc60b/gateway/webhook/webhook.go#L348)

Isn't that sufficient for now? In worse case scenario, if the store fails in first attempt, we log the error and continue. Server is not interrupted in any way.